Fix incorrect parsing of large values in VGA scratch register.

osenft opened the pull request:
The largest value in the 32 bit VGA scratch register is 0xffffffff, which is 4294967295 decimal. This requires a parsing buffer for 10 digits + `\0` + `\n`. This change increases the parsing buffer to char[12] and explicitly adds `\0` after the read string.

This fixes issue #1.

osenft committed:
Fix incorrect parsing of large values in VGA scratch register.

The largest value in the 32 bit VGA scratch register is 0xffffffff,
which is 4294967295 decimal. This requires a parsing buffer for 10
digits + \0. This change increases the parsing buffer to char[11].

Signed-off-by: Oskar Senft <osk@google.com>

jk-ozlabs commented:
Hi Oskar - thanks for the fix!

This definitely looks like an issue with the current code, but I'd suggest a tweak to the fix: the aspeed bmc-misc driver formats the register values as `"%u\n"`, so we'll need to account for the newline too, by bumping that up to char[12].

And for extra safety, we should explicitly null-terminate the string (ie, `buf[rc] = '\0'`). This way we can prevent reading past the end of `buf` in the `strtoul`. This'll only happen if the sysfs output is somehow invalid; but still worthwhile.

If you're OK to make these changes, fee free to update the PR. Or if you prefer, I can add a separate patch.

osenft committed:
Implement review comments.

Add space for a newline character and explicitly set the char after
the last read char to \0.

Signed-off-by: Oskar Senft <osk@google.com>

osenft commented:
No problem, thanks for the quick review. Good point re the `\n`, I added that as well as explicitly setting `\0`. One wonders: how did this ever work? It surely did for me for 2 years!

PTAL

jk-ozlabs commented:
> One wonders: how did this ever work? It surely did for me for 2 years!

If the sysfs file read gives us a number < 9999999, then it'll work; the strtoul will terminate on the trailing newline, so we'll get a valid parse. The VGA password value is the lower 8 bits of the register, so any value there is OK - it's just the values in the top two bytes of this 32-bit register that could cause issues with the current code. Seeing as those bytes are arbitrary scratch register contents, this may be why we've not seen a problem elsewhere.

Your fix is almost there: you'll need to subtract one from the len argument of `read()` (ie. `sizeof(buf) - 1`) otherwise your trailing-nul assignment could write one byte past the end of `buf` (ie., if read returns `sizeof(buf)`)

osenft committed:
Fixed max buffer size for read.

Don't allow the read to write to the last byte in the buffer, since
that's reserved for the \0 char.

Signed-off-by: Oskar Senft <osk@google.com>

osenft commented:
Argh, yeah, sorry for that. I should have considered that. I also updated the comment to make it clearer that it includes the `\0`.

PTAL. Hopefully I got it right this time.

jk-ozlabs commented:
All looks great, thanks for the updates. Want to squash these into a single commit that implements the complete fix?

(or I can do that, if you prefer)

osenft commented:
Thanks! Doesn't Github have the option to squash + merge?

jk-ozlabs commented:
Merged, thanks!

Github can do the squash & merge, but doing it manually allows you to define the resulting commit message.
